### PR TITLE
support multiple region selection in code snippet

### PR DESCRIPTION
**Pull request by superyyrrzz**
support #189 Part1 @vicancy

### CODE REVIEW

File: Documentation/spec/docfx_flavored_markdown.md [hunk 1]
```diff
@@ -111,7 +111,7 @@ Allows you to insert code with code language specified. The content of specified
 * __`<codepath>`__ is the relative path in file system which indicates the code snippet file that you want to expand.
 * __`<queryoption>`__ and __`<queryoptionvalue>`__ are used together to retrieve part of the code snippet file in the line range or tag name way. We have 2 query string options to represent these two ways:
```
Comment (vicancy) [inline Documentation/spec/docfx_flavored_markdown.md]: 3 query string options?

File: src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs [hunk 2]
```diff
@@ -233,42 +233,81 @@ public DfmExtractCodeResult ExtractFencesCode(DfmFencesBlockToken token, string
...
+
+            return new DfmExtractCodeResult
+            {
+                IsSuccessful = true,
+                FencesCodeLines = TrimIndent(includedLines, indentLength),
+            };
+        }
+
+        private void AddFencesCodeRegion(List<string> includedLines, string[] codeLines, int startLine, int endLine, int indentLength, HashSet<int> excludedLines = null)
+        {
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs]: indentLength not used?

File: src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs [hunk 3]
```diff
@@ -233,42 +233,81 @@ public DfmExtractCodeResult ExtractFencesCode(DfmFencesBlockToken token, string
...
                 }
             }
+        }
 
-            return new DfmExtractCodeResult
-            {
-                IsSuccessful = true,
-                FencesCodeLines = (indentLength == 0 ? includedLines : includedLines.Select(s => Regex.Replace(s, string.Format(RemoveIndentSpacesRegexString, indentLength), string.Empty))).ToArray()
-            };
+        private string[] TrimIndent(List<string> codeLines, int indentLength)
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs]: what is the indent trim principle? *Collect all the ranges and then calculate the indent together* or *calculate indent for each region and than collect the trimmed ranges*?

File: Documentation/spec/docfx_flavored_markdown.md (hunk 1 above)
Comment (superyyrrzz) [inline Documentation/spec/docfx_flavored_markdown.md]: 2 option: `#` and `?`, and option`?` has 3 sub options ....
I think it's more clear if I use a table.

File: src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs (hunk 3 above)
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs]: Collect all the ranges and then calculate the indent together. I want to preserve the relative indent between regions.

File: src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs (hunk 2 above)
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs]: nice catch... will remove it

File: src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs [hunk 4]
```diff
@@ -233,42 +233,81 @@ public DfmExtractCodeResult ExtractFencesCode(DfmFencesBlockToken token, string
 
                 return GetFencesCodeCore(fencesCode, resolveResult.StartLine, resolveResult.EndLine, resolveResult.IndentLength, resolveResult.ExcludesLines);
             }
-            else
+            else if (token.PathQueryOption?.Regions != null)
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs]: abstract start-end, range to a method returns included lines with type IEnumerable<int>?

Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs]: got your idea... will update a new version.

File: Documentation/spec/docfx_flavored_markdown.md [hunk 5]
```diff
@@ -84,13 +84,13 @@ Block file inclusion must be in a single line and with no prefix characters befo
...
 Give an example below.
 
-    > [!div class="tabbedCodeSnippets" data-resources="OutlookServices.Calendar"]
-    > ```cs
-    > <cs code text>
-    > ```
-    > ```javascript
-    > <js code text>
-    > ```
+    [!div class="tabbedCodeSnippets" data-resources="OutlookServices.Calendar"]
```
Comment (vicancy) [inline Documentation/spec/docfx_flavored_markdown.md]: div should be in blocks? why removing `>`

**Comment by vicancy** [comment]
:+1:

**Comment by vwxyzh** [comment]
👍

**Pull request closed without merging**